lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,850 @@
+            certificate_data[key] = raw_relation_data[key]
+    return certificate_data
+
+
+def generate_private_key(
+    password: Optional[bytes] = None,
+    key_size: int = 2048,
+    public_exponent: int = 65537,
+) -> bytes:
+    """Generates a private key.
+
+    Args:
+        password (bytes): Password for decrypting the private key
+        key_size (int): Key size in bytes
+        public_exponent: Public exponent.
+
+    Returns:
+        bytes: Private Key
+    """
+    private_key = rsa.generate_private_key(
+        public_exponent=public_exponent,
+        key_size=key_size,
+    )
+    key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.BestAvailableEncryption(password)
+        if password
+        else serialization.NoEncryption(),
+    )
+    return key_bytes
+
+
+def generate_csr(
+    private_key: bytes,
+    subject: str,
+    private_key_password: Optional[bytes] = None,
+    sans: Optional[List[str]] = None,
+    additional_critical_extensions: Optional[List] = None,
+) -> bytes:
+    """Generates a CSR using private key and subject.
+
+    Args:
+        private_key (bytes): Private key
+        private_key_password (bytes): Private key password
+        subject (str): CSR Subject.
+        sans (list): List of subject alternative names
+        additional_critical_extensions (list): List if critical additional extension objects.
+            Object must be a x509 ExtensionType.
+
+    Returns:
+        bytes: CSR
+    """
+    signing_key = serialization.load_pem_private_key(private_key, password=private_key_password)
+    csr = x509.CertificateSigningRequestBuilder(
+        subject_name=x509.Name(
+            [
+                x509.NameAttribute(x509.NameOID.COMMON_NAME, subject),
+            ]
+        )
+    )
+    if sans:
+        csr = csr.add_extension(
+            x509.SubjectAlternativeName([x509.DNSName(san) for san in sans]), critical=False
+        )
+    if additional_critical_extensions:
+        for extension in additional_critical_extensions:
+            csr = csr.add_extension(extension, critical=True)
+    signed_certificate = csr.sign(signing_key, hashes.SHA256())  # type: ignore[arg-type]
+    return signed_certificate.public_bytes(serialization.Encoding.PEM)
+
+
+class CertificatesProviderCharmEvents(CharmEvents):
+    """List of events that the TLS Certificates provider charm can leverage."""
+
+    certificate_request = EventSource(CertificateRequestEvent)
+    certificate_revokal = EventSource(CertificateRevokalEvent)
+
+
+class CertificatesRequirerCharmEvents(CharmEvents):
+    """List of events that the TLS Certificates requirer charm can leverage."""
+
+    certificate_available = EventSource(CertificateAvailableEvent)
+    certificate_expiring = EventSource(CertificateExpiringEvent)
+    certificate_expired = EventSource(CertificateExpiredEvent)
```

Comment:
Maybe we should add the revocation event here also, because there are reasons to have the provider initiate the revocation (https://datatracker.ietf.org/doc/html/rfc5280#section-5.3.1)

Reply:
Done